portal-api/src/v1/users/sign-in-link.controller.login-with-sign-in-link.test.js

```diff
@@ -0,0 +1,169 @@
+        });
+
+        it('should respond with a 200', async () => {
+          await signInLinkController.loginWithSignInLink(req, res);
+
+          expect(res.status).toHaveBeenCalledWith(200);
+        });
+
+        it('should respond with the login token and user details', async () => {
+          await signInLinkController.loginWithSignInLink(req, res);
+
+          expect(res.json).toHaveBeenCalledWith(successfulLoginResponse);
+        });
+      });
+
+      describe('given loginUser throws an error', () => {
+        const loginUserError = new Error('test error');
+
+        beforeEach(() => {
+          when(signInLinkService.loginUser)
+            .calledWith(TEST_USER._id)
+            .mockRejectedValueOnce(loginUserError);
+        });
+
+        itShouldReturnA500WithMessage(loginUserError.message);
+      });
+    });
+
+    describe('given isValidSignInToken returns false', () => {
+      beforeEach(() => {
+        mockSuccessfulIsValidSignInTokenReturnFalse();
+      });
+
+      it('should not call deleteSignInToken on the signInLinkService', async () => {
+        await signInLinkController.loginWithSignInLink(req, res);
+
+        expect(signInLinkService.deleteSignInToken).not.toHaveBeenCalled();
+      });
+
+      itShouldReturnA403();
+    });
+
+    describe('given isValidSignInToken throws an InvalidSignInTokenError', () => {
+      beforeEach(() => {
+        mockUnsuccessfulIsValidSignInToken();
+      });
+
+      itShouldReturnA403();
+    });
+
+    function itShouldReturnA403() {
+      it('should respond with a 403', async () => {
+        await signInLinkController.loginWithSignInLink(req, res);
+
+        expect(res.status).toHaveBeenCalledWith(403);
+
+        expect(res.send).toHaveBeenCalledWith({
+          message: 'Forbidden',
+          errors: [{
+            msg: `Invalid sign in token for user ID: ${TEST_USER._id}`,
+          }],
+        });
+      });
+    }
+
+    function itShouldReturnA500WithMessage(message) {
+      it('should respond with a 500', async () => {
+        await signInLinkController.loginWithSignInLink(req, res);
+
+        expect(res.status).toHaveBeenCalledWith(500);
+      });
+
+      it('should respond with expected error message', async () => {
+        await signInLinkController.loginWithSignInLink(req, res);
+
+        expect(res.send).toHaveBeenCalledWith({
+          message: 'Internal Server Error',
+          errors: [{
+            msg: message,
+          }],
+        });
+      });
+    }
+
+    function mockUnsuccessfulIsValidSignInToken() {
+      when(signInLinkService.isValidSignInToken).calledWith(expect.anything()).mockRejectedValue(new InvalidSignInTokenError(TEST_USER._id));
+    }
+
+    function mockSuccessfulIsValidSignInToken(resolvedValue) {
+      when(signInLinkService.isValidSignInToken).calledWith(expect.anything()).mockResolvedValue(resolvedValue);
```

Comment:
As far as I can tell, we don't check that we actually call `isValidSignInToken` with the expected arguments at any point in these tests. I think we should either change this function to not use `expect.anything()`, or add a test case to check that we call it with the correct arguments.

Reply:
Good point, I've adjusted the mock to check the input